### DISCUSSION TITLE
Design Idea: Settlement Templates

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#450] [Ideas]
Most program planners agree that the footprint of the first generation of settlements on Mars should be `cylindrical`.

This is because one has to fit various modules such as habitation, service. power or lab, etc. within the circular shell of a rocket. 

To build the first colony, we will need automate guidance control to direct these modules to arrive on a particular landing zone. Upon safe landing, they will automatically link up and initialize their systems to do diagnostics in preparation for the arrival of humans. 

No doubt, a `building block` construction approach will be touted as being the most feasible and cost-effective. 

At some points in future as we learn to erect structures safely and efficiently on Mars, the construction of giant, semi-inflatable dome won't be far off. 

In recent years, numerous research have focused on how to 3-D print self-contained housing units on Mars that rely on in-situ material made from Mars Regolith.

In mars-sim, perhaps it's time we begin to explore on how to program different kinds of settlement layouts like the one below : 

![Mars colony layout a](https://user-images.githubusercontent.com/1168584/62668968-a3e22e00-b942-11e9-95b0-bdfaa85b85df.jpg)

Let's brainstorm ways to upgrade mars-sim's core engine so as to construct settlements to look different from the `cookie-cutter` or `modular` approach that most strategy games rely on.

Feel free to comment and provide feedback.

**Comment by Urwumpe** [comment]
Well, that looks like architecture, maybe we should treat it like that: Define rooms from basic geometric shapes, intersections and subdivisions.  For example, starting out with a basic circle and then dividing this into rooms by creating rings and replacing the rings inside the circle then by ring segment rooms... Of course, this means such a building would be that: One complex shaped building. Currently we have multiple buildings with dedicated functions.

**Comment by mokun** [comment]
> Define rooms from basic geometric shapes, intersections and subdivisions. 

Yay If we could come up with an interface for users to draw out a building with some basic shapes, it would be fun !

**Comment by Urwumpe** [comment]
Sure, but not easy. I think the modular approach is still better for getting ahead, since we would need some kind of 2D CAD tool to let players design buildings or view designs by AI.

**Comment by mokun** [comment]
Yay we'll need to first develop an layout editor in order to open up the freedom of designing the settlement layout. 

For now, we're stuck with the modular concept and there are still a lot of things that we may improve.

For instance, we have recently added a new separate building called "EVA Airlock" and removed the EVA function from the Lander Hab in [5178](https://github.com/mars-sim/mars-sim/commit/991727e60114ffc44587f447a8098a8cef4ecac6). 

The reason for this change is that we have encountered an occasional bug as settlers are inside the airlock of a Lander Hab. 

Separating the EVA into its own building/room may help isolating the root cause for future debugging.

Besides, the Lander Hab is really very very full.

No airlock room is drawn in Lander Hab. 

The location where the settlers begin to do the EVA is actually inside the gym area.

**Comment by mokun** [comment]
@Urwumpe @bevans2000 ,

I'm expanding the so-called *Settlement Template* in settlements.xml.

In reality, at least for now, there are just phases of a building a settlement, such as *Mars Direct Base (Phase 2), Alpha Base (Phase 4), etc.* .

I can see that each sponsor (or space agency) should follow its own path of development, rather than demanding them necessarily adopt the *Mars Direct Plan*.

So, I'll retain the *Mars Direct Plan* in mars-sim as a way of paying tribute to the Mars Society.

But the rest of the sponsors should be FREE to adopt their own methodology and system architecture for developing a colony on Mars.

I know this is just one small step toward the right direction in modeling since it reflects better what the reality is.

And we can readily imagine how difficult SpaceX as a business enterprise would do things very differently from a government led human settlement program on Mars.

**Comment by mokun** [comment]
With various phases of development made possible, beyond that of the Mars Direct Base plan, we could explore having totally different building layouts. 

Here are some of the inspirational art work.
A. Central Dome
![central dome 2](https://user-images.githubusercontent.com/1168584/121569911-2b34d880-c9d6-11eb-8fd3-f1c7a311318b.jpg)

B. Hyperloop on Mars with Cliff Dwelling 
![hyperloop on Mars and Cliff dwelling](https://user-images.githubusercontent.com/1168584/121570022-4c95c480-c9d6-11eb-94d7-0e431c14ae55.jpg)

C. Distributed Domes and Round Transportation Terminal 
![round terminal](https://user-images.githubusercontent.com/1168584/121570118-69ca9300-c9d6-11eb-81b4-fd28173a367e.jpg)

**Comment by mokun** [comment]
@Urwumpe @bevans2000 

Here's another aspirational artwork that help visualizing how linking modules together may result in a basic and functional human colony.

![image](https://user-images.githubusercontent.com/1168584/173109194-cd5cbad4-a736-4a1f-8b68-7fc06085b559.png)

**Comment by mokun** [comment]
@Urwumpe @bevans2000 

As we have been discussing the prospect of having a *Project* class, I'd like to connect it to the important ability for a player to choose and dictate how he like to expand his settlement.

e.g. Say, a player may select from a customizable template of constructing a *East Wing* to the existing settlement.

The blueprint of this *East Wing* may include 2 Residential Buildings, 1 Laboratory, 2 Storage Sheds, 1 Inflatable Greenhouse, etc.

Once the player chooses this construction template, the countdown would start, the whole settlement will work toward this goal. 

Also, we could make use of the idea of *private enterprise sponsorship* of these types of ***projects*** to push boundary.

Say, we may even offer *reward* if the second settlement choose to join in this competition.

When both settlement complete a standard template project and the results are compared, whoever sponsors this project would reward the winning settlement with a higher publicity or attractiveness score, which would in terms attract more talents.

**Reply by Urwumpe** [reply]
I think this makes sense, but since we are in hostile space, I would include the whole base as scope for the change and make sure, the new buildings won't strain the available resources.

**Comment by mokun** [comment]
For now, our settlement design is more or less derived from a cylindrical portion of a rocket (or a rectangular module that can be fitted inside a rocket outer shell). 

In future, imagine we have the following types of settlements that are locale based: 
1. Open flat design on a plain (what we have currently)
2. Embedded in cliff-side, hillside or crater rim
3. Inside canyon or chasm
4. Inside lava tube
5. Embedded on a gullies

**Reply by bevans2000** [reply]
The first thing we need to focus on is making it easier/simper to build Buildings. At the moment there are a few steps, create the base, build the shell etc etc. I think we need to consider how the new ```Project``` concept could help my creating a Project Plan to do all the various stages. Having that flexible concept would help. Certainly it need to be tackled in 3.5.0.

**Reply by mokun** [reply]
> making it easier/simper to build Buildings.

Yes. I see we'll need to open a new project to tackle on how to make it a breeze to create new buildings in order to expand the settlement.

**Reply by mokun** [reply]
Also having a graphic editor for us developer to easily create new and unique open space settlement template is a must.

Currently, it's very time-consuming to get all rotations as well as x and y params right when placing and connecting buildings together in a xml settlement template.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

I just finished the 1st round in designing a 20m x 20m central dome building at the latest [commit](https://github.com/mars-sim/mars-sim/commit/b3906663e5408702002ca390a16d3e291b242288).

![image](https://github.com/mars-sim/mars-sim/assets/1168584/10dd57a7-66d0-4354-b8ed-14a0f978b0d9)

I just need to touch up with better color and rework all activity spots.

What other buildings would you like to add to this new brand new settlement template called "Hub Base" ?

**Reply by Urwumpe** [reply]
What about a recreation/sports dome?

A hospital might also be more ergonomic in a dome. 

Or a larger habitat with multiple Apartments inside? And of course: what about using the basic Dome idea for a large water tank?

**Reply by bevans2000** [reply]
Looks good.

**Reply by mokun** [reply]
The Hub base now finally has East Wing, South Wing, and West Wings : 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/2c8cd960-e9dd-4546-a525-37c7fa914c2d)

**Reply by mokun** [reply]
> What about a recreation/sports dome?

Yay we can create one in near future as we scale up mars-sim.

> A hospital might also be more ergonomic in a dome.

Hospital architecture is supposed to be on the conservative side because it should emphasize utility value, not aesthetic :D 

> Or a larger habitat with multiple Apartments inside? 

If we want to create apartment buildings, then we'll need multi-floor layouts. 

We'll have to change the walk-related class. It could be a nightmare but about time to look into them. 

> And of course: what about using the basic Dome idea for a large water tank?

- Right. A water tower creates water pressure and we haven't considered pipingd and electrical connections just yet. So we'll need services from plumbers soon :)
